pkg/query/listener.go

```diff
@@ -0,0 +1,144 @@
+}
+
+// ExitUnivariate is called when production univariate is exited.
+func (s *queryListener) ExitUnivariate(ctx *parser.UnivariateContext) {
+	if s.err != nil {
+		return
+	}
+	leftOp, operator, rightOp := getCriterionFields(ctx.Key(), ctx.UniOp(), ctx.Value())
+	s.leftOp = leftOp
+	s.op = operator
+	s.rightOp = []string{rightOp}
+	s.err = s.storeCriterion()
+}
+
+// ExitManyValues is called when production manyValues is exited.
+func (s *queryListener) ExitManyValues(ctx *parser.ManyValuesContext) {
+	if s.err != nil {
+		return
+	}
+	_, _, rightOp := getCriterionFields(nil, nil, ctx.Value())
+	s.rightOp = append(s.rightOp, rightOp)
+}
+
+// ExitMultivariate is called when production univariate is exited.
+func (s *queryListener) ExitMultivariate(ctx *parser.MultivariateContext) {
+	if s.err != nil {
+		return
+	}
+
+	leftOp, operator, _ := getCriterionFields(ctx.Key(), ctx.MultiOp(), nil)
+	s.leftOp = leftOp
+	s.op = operator
+	if s.rightOp == nil {
+		s.rightOp = []string{""}
+	}
+	s.err = s.storeCriterion()
+}
+
+func (s *queryListener) storeCriterion() error {
+	operator, err := findOpByString(s.op)
+	if err != nil {
+		return err
+	}
+	criterion := NewCriterion(s.leftOp, operator, s.rightOp, s.criteriaType)
+	if err := criterionDateToUTC(&criterion); err != nil {
+		return err
+	}
+	if err = criterion.Validate(); err != nil {
+		return err
+	}
+	s.result = append(s.result, criterion)
+	return nil
+}
+
+func (s *queryListener) ReportAmbiguity(recognizer antlr.Parser, dfa *antlr.DFA, startIndex, stopIndex int, exact bool, ambigAlts *antlr.BitSet, configs antlr.ATNConfigSet) {
+}
+
+func (s *queryListener) ReportAttemptingFullContext(recognizer antlr.Parser, dfa *antlr.DFA, startIndex, stopIndex int, conflictingAlts *antlr.BitSet, configs antlr.ATNConfigSet) {
+}
+
+func (s *queryListener) ReportContextSensitivity(recognizer antlr.Parser, dfa *antlr.DFA, startIndex, stopIndex, prediction int, configs antlr.ATNConfigSet) {
+}
+
+func (s *queryListener) SyntaxError(recognizer antlr.Recognizer, offendingSymbol interface{}, line, column int, msg string, e antlr.RecognitionException) {
+	s.err = &util.UnsupportedQueryError{Message: fmt.Sprintf("error while parsing %s at column %d: %s", s.criteriaType, column, msg)}
+}
+
+func getCriterionFields(key, op, right antlr.TerminalNode) (leftOp string, operator string, rightOp string) {
+	if key != nil {
+		leftOp = key.GetText()
+	}
+	if right != nil {
+		rightOp = right.GetText()
+		rightOp = strings.Replace(rightOp, "''", "'", -1)
+		rightOp = strings.Trim(rightOp, "'")
+	}
+	if op != nil {
+		operator = op.GetText()
+	}
+	return
+}
+
+func criterionDateToUTC(c *Criterion) error {
+	if c.LeftOp != "created_at" && c.LeftOp != "updated_at" {
+		return nil
+	}
+	rightOp := make([]string, 0, len(c.RightOp))
+	for _, r := range c.RightOp {
+		parsedTime, err := time.Parse(time.RFC3339Nano, r)
+		if err != nil {
+			return err
+		}
+		formatted := parsedTime.UTC().Format(time.RFC3339Nano)
```

Comment:
This is for testing currently. I haven't decided how to address the format change, because it's also a breaking change in the code - maybe introduce a `FormatRFCNano` method and deprecate the old `FormatRFC`